internal/webhooks/patch_validation.go

```diff
@@ -0,0 +1,360 @@
+	if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Variable != nil {
+		// If the variable is one of the list of builtin variables it's valid.
+		if strings.HasPrefix(*jsonPatch.ValueFrom.Variable, "builtin") {
+			if _, ok := builtinVariables[*jsonPatch.ValueFrom.Variable]; !ok {
+				allErrs = append(allErrs,
+					field.Invalid(
+						path.Child("valueFrom", "variable"),
+						jsonPatch.ValueFrom.Variable,
+						"not a defined builtin variable",
+					))
+			}
+		} else {
+			if _, ok := variableSet[*jsonPatch.ValueFrom.Variable]; !ok {
+				allErrs = append(allErrs,
+					field.Invalid(
+						path.Child("valueFrom", "variable"),
+						jsonPatch.ValueFrom.Variable,
+						fmt.Sprintf("variable with name %s cannot be found", *jsonPatch.ValueFrom.Variable),
+					))
+			}
+		}
+	}
+	return allErrs
+}
+
+// This contains a list of all of the valid builtin variables.
+// TODO(killianmuldoon): Match this list to controllers/topology/internal/extensions/patches/variables as those structs become available across the code base i.e. public or top-level internal.
+var builtinVariables = sets.NewString(
+	"builtin",
+
+	// Cluster builtins.
+	"builtin.cluster",
+	"builtin.cluster.name",
+	"builtin.cluster.namespace",
+
+	// ClusterTopology builtins.
+	"builtin.cluster.topology",
+	"builtin.cluster.topology.version",
+	"builtin.cluster.topology.class",
+
+	// ControlPlane builtins.
+	"builtin.controlPlane",
+	"builtin.controlPlane.version",
+	"builtin.controlPlane.replicas",
+
+	// MachineDeployment builtins.
+	"builtin.machineDeployment",
+	"builtin.machineDeployment.version",
+	"builtin.machineDeployment.class",
+	"builtin.machineDeployment.name",
+	"builtin.machineDeployment.topologyName",
+	"builtin.machineDeployment.replicas",
+)
+
+func getClusterClassVariablesMap(clusterClassVariables []clusterv1.ClusterClassVariable) map[string]*clusterv1.ClusterClassVariable {
+	variablesMap := map[string]*clusterv1.ClusterClassVariable{}
+	for i := range clusterClassVariables {
+		variablesMap[clusterClassVariables[i].Name] = &clusterClassVariables[i]
+	}
+	return variablesMap
+}
+
+// validateIndexAccess checks to see if the jsonPath is attempting to add an element in the array i.e. access by number
+// If the operation is add an error is thrown if a number greater than 0 is used as an index.
+// If the operation is replace an error is thrown if an index is used.
+func validateIndexAccess(jsonPatch clusterv1.JSONPatch, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	pathParts := strings.Split(jsonPatch.Path, "/")
+	for _, part := range pathParts {
+		// If the jsonPatch operation is add disallow any number greater than 0. 0 here is a prepend operation on the array.
+		index, err := strconv.Atoi(part)
+		if err != nil {
+			continue
+		}
+
+		if jsonPatch.Op == "add" && index != 0 {
+			allErrs = append(allErrs,
+				field.Invalid(path,
+					jsonPatch.Path,
+					"arrays can only be accessed using [0] (prepend) or [-] (append)",
```

Comment:
```suggestion
					"it's only possible to add to arrays using \"/0\" (prepend) or \"/-\" (append)",
```
nit, maybe it's better as it's the valid syntax, otherwise users might think they can address it via literally `[0]`